dids/src/main/kotlin/web5/sdk/dids/DidDht.kt

```diff
@@ -0,0 +1,420 @@
+
+    /**
+     * Creates a new "did:dht" DID, derived from an initial identity key, and stores the associated private key in the
+     * provided [KeyManager].
+     *
+     * The method-specific identifier of a "did:dht" DID is a z-base-32 encoded public key.
+     *
+     * **Note**: By default, no additional keys or services are added to the document
+     *
+     * @param keyManager A [KeyManager] instance where the new key will be stored.
+     * @param options Optional parameters ([CreateDidDhtOptions]) to specify additional keys, services, and optional
+     * publishing during creation.
+     * @return A [DidDht] instance representing the newly created "did:dht" DID.
+     */
+    override fun create(keyManager: KeyManager, options: CreateDidDhtOptions?): DidDht {
+      // TODO(gabe): enforce that provided keys are of supported types according to the did:dht spec
+      val opts = options ?: CreateDidDhtOptions()
+
+      // create identity key
+      val keyAlias = keyManager.generatePrivateKey(JWSAlgorithm.EdDSA, Curve.Ed25519)
+      val publicKey = keyManager.getPublicKey(keyAlias)
+
+      // build DID Document
+      val id = getDidIdentifier(publicKey)
+
+      // add identity key to relationships map
+      val identityVerificationMethod =
+        VerificationMethod.builder()
+          .id(URI.create("$id#0"))
+          .type("JsonWebKey2020")
+          .controller(URI.create(id))
+          .publicKeyJwk(publicKey.toPublicJWK().toJSONObject())
+          .build()
+
+      // add all other keys to the verificationMethod and relationships arrays
+      val relationshipsMap = mutableMapOf<PublicKeyPurpose, MutableList<VerificationMethod>>().apply {
+        val identityVerificationMethodRef = VerificationMethod.builder().id(identityVerificationMethod.id).build()
+        listOf(
+          PublicKeyPurpose.AUTHENTICATION,
+          PublicKeyPurpose.ASSERTION_METHOD,
+          PublicKeyPurpose.CAPABILITY_DELEGATION,
+          PublicKeyPurpose.CAPABILITY_INVOCATION
+        ).forEach { purpose ->
+          getOrPut(purpose) { mutableListOf() }.add(identityVerificationMethodRef)
+        }
+      }
+
+      // map to the DID object model's verification methods
+      val verificationMethods = (opts.verificationMethodsToAdd?.map { (key, purposes) ->
+        VerificationMethod.builder()
+          .id(URI.create("$id#${key.keyID}")).type("JsonWebKey2020")
+          .controller(URI.create(id))
+          .publicKeyJwk(key.toPublicJWK().toJSONObject())
+          .build().also { verificationMethod ->
+          purposes.forEach { relationship ->
+            relationshipsMap.getOrPut(relationship) { mutableListOf() }.add(
+              VerificationMethod.builder().id(verificationMethod.id).build()
+            )
+          }
+        }
+      } ?: emptyList()) + identityVerificationMethod
+
+      // map to the DID object model's services
+      val services = opts.servicesToAdd?.map { service ->
```

Comment:
Should we have a require that makes sure all parts of the servicesToAdd are there?
(What if we have a null type or service endpoint?)